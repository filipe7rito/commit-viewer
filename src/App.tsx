import React from 'react';

function App() {
  return <h1>Hello Codacy</h1>;
}

export default App;
