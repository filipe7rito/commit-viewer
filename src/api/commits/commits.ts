import axios from 'axios';
import { axiosRequestConfig } from '../apiConfig';
import { Commit } from '../../types/commit';

/**
 * @param {username} username -github username
 * @param {repository} repository - github repository
 * @param {page} page - commits api page number
 * @param {pageSize} pageSize - commits pages size
 *
 * @returns list of commits for a given repository
 */
export async function fetch({
  username,
  repository,
  page,
  pageSize,
}: {
  username: string;
  repository: string;
  page?: number;
  pageSize?: number;
}): Promise<Commit[]> {
  const response = await axios.get(`/repos/${username}/${repository}/commits`, {
    ...axiosRequestConfig,
    params: {
      per_page: pageSize,
      page,
    },
  });

  const commits: Commit[] = response.data.map((record: any) => {
    return {
      sha: record.sha,
      message: record.commit.message,
      author: {
        id: record.author ? record.author.id : record.committer.id,
        login: record.author ? record.author.login : record.committer.login,
        name: record.commit.author.name,
        avatarUrl: record.author ? record.author.avatar_url : record.committer.avatar_url,
        date: record.commit.author.date,
      },
      parents: record.parents,
    };
  });

  return commits;
}
