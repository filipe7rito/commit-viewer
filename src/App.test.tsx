import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import 'intersection-observer';
import { rest } from 'msw';
import React from 'react';
import App from './App';
import { commitsResponse } from './test/fixtures/commits';
import { baseUrl } from './test/mocks/handlers';
import { server } from './test/mocks/server';

describe('App', () => {
  it('Renders app', async () => {
    const { asFragment, getByText } = renderApp();

    await waitFor(() => {
      getByText(/your commits will show up here/i);
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('Search url and list commits', async () => {
    const { getByLabelText, getByTestId, getByText } = renderApp();

    userEvent.type(
      getByLabelText(/repository url/i),
      'https://github.com/amazing-user/amazing-repo',
    );

    userEvent.click(getByText(/load commits/i));

    await waitFor(() => {
      // First row result in table
      expect(getByTestId('table-row-0')).toBeInTheDocument();
    });
  });

  // This test throws an error on setState due to IntersectionObserver
  it('Loads more commits on scroll', async () => {
    const { getByLabelText, getByText, getByTestId, getAllByRole } = renderApp();

    userEvent.type(
      getByLabelText(/repository url/i),
      'https://github.com/amazing-user/amazing-repo',
    );

    userEvent.click(getByText(/load commits/i));

    await waitFor(() => {
      // First row result in table
      expect(getByTestId('table-row-0')).toBeInTheDocument();
    });

    const tableBody = getByTestId('table-body');
    const lastRow = getByTestId(`table-row-${commitsResponse.length - 1}`);

    // Scroll to lastRow
    fireEvent.scroll(tableBody, lastRow);

    await waitFor(() => {
      // Header row plus body rows
      expect(getAllByRole('row').length).toEqual(21);
      expect(getByTestId('table-row-14')).toBeInTheDocument();
    });
  });

  it('Filters commits list', async () => {
    const { getByLabelText, getByTestId, getByText, getAllByRole } = renderApp();

    userEvent.type(
      getByLabelText(/repository url/i),
      'https://github.com/amazing-user/amazing-repo',
    );

    userEvent.click(getByText(/load commits/i));

    await waitFor(() => {
      // First row result in table
      expect(getByTestId('table-row-0')).toBeInTheDocument();
    });

    userEvent.type(getByLabelText(/search/i), commitsResponse[0].sha);

    // Header row plus only row filtered
    expect(getAllByRole('row').length).toEqual(2);
  });

  it('Search an invalid url', async () => {
    const { getByLabelText, getByText } = renderApp();

    userEvent.type(getByLabelText(/repository url/i), 'an-invalid-url');

    userEvent.click(getByText(/load commits/i));

    expect(screen.getByText(/this url is not valid/i)).toBeInTheDocument();
  });

  it('Fails getting commits', async () => {
    const { getByLabelText, getByText } = renderApp();

    userEvent.type(
      getByLabelText(/repository url/i),
      'https://github.com/amazing-user/amazing-repo',
    );

    userEvent.click(getByText(/load commits/i));

    server.use(
      rest.get(`${baseUrl}/repos/*/commits`, (_req, res, ctx) => {
        return res(ctx.status(500));
      }),
    );

    await waitFor(() => {
      expect(
        screen.getByText(/we couldnâ€™t get the commits for this repository/i),
      ).toBeInTheDocument();
    });
  });

  function renderApp() {
    return render(<App />);
  }
});
