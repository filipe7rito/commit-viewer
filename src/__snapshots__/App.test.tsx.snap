// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Renders app 1`] = `
<DocumentFragment>
  <div
    class="css-8sdul9"
  >
    <div
      class="css-11e6jbq"
    >
      <nav
        class="css-d04ge1"
      >
        <a
          href="/"
        >
          <div
            class="css-wtwgib"
          >
            <img
              alt=""
              src="https://app.codacy.com/static/media/codacy-logo.be9caad5.svg"
            />
          </div>
        </a>
      </nav>
      <nav
        class="css-zy7d5y"
      >
        <form>
          <div
            class="row g-3 align-items-baseline"
          >
            <div
              class="col-auto"
            >
              <label
                class="col-form-label"
                for="search-input"
              >
                Repository URL
              </label>
            </div>
            <div
              class="input-group-sm col-auto position-relative css-5guh9b"
            >
              <input
                aria-describedby="search-input"
                autocomplete="off"
                class="form-control css-190wxqs"
                id="search-input"
                type="text"
                value=""
              />
              <div
                class="invalid-tooltip"
              >
                This URL is not valid
              </div>
            </div>
            <div
              class="col-auto mt-1"
            >
              <button
                class="btn btn-primary btn-sm css-svn9b6"
                type="submit"
              >
                Load commits
              </button>
            </div>
          </div>
        </form>
      </nav>
    </div>
    <main
      class="css-1mhygto"
    >
      <aside>
        <ul
          class="nav flex-column css-1upl4pb"
        >
          <li>
            <img
              alt="back-arrow"
              src="https://app.codacy.com/legacy/versioned/images/navigation/arrow-back.svg"
            />
            Team
            <hr />
          </li>
          <li>
            <svg
              viewBox="0 0 50 35"
            >
              <path
                d="M36.2,24.95L29.94,9.62a0.29,0.29,0,0,0-.48,0l-2.1,4.8a0.22,0.22,0,0,1-.33.08l-2-1.55a0.21,0.21,0,0,0-.31.07L21.7,19.66a0.2,0.2,0,0,1-.36,0l-2.7-5.84a0.2,0.2,0,0,0-.37,0l-2.51,8.43a0.2,0.2,0,0,1-.26.13l-2-.78v2.14L16.79,25a0.2,0.2,0,0,0,.26-0.13l1.61-5.42a0.2,0.2,0,0,1,.37,0l2.3,5a0.2,0.2,0,0,0,.36,0l3.8-8.27a0.2,0.2,0,0,1,.31-0.07l2,1.58a0.22,0.22,0,0,0,.34-0.09l1.13-2.66a0.21,0.21,0,0,1,.39,0l4,10h2.6A0.27,0.27,0,0,1,36.2,24.95Z"
              />
              <path
                d="M34.11,27H13.2c-0.11,0-.08.09-0.08,0.2v1.6c0,0.11,0,.2.08,0.2H37.12a17.76,17.76,0,0,0-.32-2H34.11Z"
              />
            </svg>
            Dashboard
          </li>
          <li
            class="css-gzf0p2"
          >
            <svg
              class="svg aside-icon replaced-svg"
              fill="#7E90B2"
              viewBox="0 0 50 35"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24.89,13.7a7.3,7.3,0,1,0,7.3,7.3A7.3,7.3,0,0,0,24.89,13.7Zm0,12.39a4.84,4.84,0,0,1-5-5.09,4.94,4.94,0,0,1,5-5.09A4.76,4.76,0,0,1,29.83,21,4.72,4.72,0,0,1,24.89,26.09Z"
              />
              <path
                d="M32.05,19.93H36.8a0.2,0.2,0,0,1,.2.2v1.6a0.2,0.2,0,0,1-.2.2H32.05a0,0,0,0,1,0,0v-2A0,0,0,0,1,32.05,19.93Z"
              />
              <path
                d="M13.2,19.93h4.75a0,0,0,0,1,0,0v2a0,0,0,0,1,0,0H13.2a0.2,0.2,0,0,1-.2-0.2v-1.6A0.2,0.2,0,0,1,13.2,19.93Z"
              />
            </svg>
            Commits
          </li>
          <li>
            <svg
              viewBox="0 0 50 35"
            >
              <path
                d="M20.56,18V17a0.38,0.38,0,0,1,.13-0.29l2.72-2.27c0.42-.35.77-0.19,0.77,0.36v0.24a0.38,0.38,0,0,1-.13.29l-2.13,1.83a0.39,0.39,0,0,0,0,.58L24,19.62a0.38,0.38,0,0,1,.13.29v0.24c0,0.55-.35.71-0.77,0.36l-2.72-2.25A0.38,0.38,0,0,1,20.56,18Z"
              />
              <path
                d="M26.37,20.51c-0.42.35-.77,0.19-0.77-0.36V19.91a0.39,0.39,0,0,1,.13-0.29l2.12-1.83a0.39,0.39,0,0,0,0-.59l-2.12-1.82a0.39,0.39,0,0,1-.13-0.29V14.85c0-.55.35-0.71,0.77-0.36l2.72,2.26a0.39,0.39,0,0,1,.13.29V18a0.39,0.39,0,0,1-.13.29Z"
              />
              <path
                d="M33.51,26.85a1,1,0,0,1-1,1h-15a1,1,0,0,1-1-1V8.15a1,1,0,0,1,1-1h15a1,1,0,0,1,1,1V26.85ZM18.28,25a1,1,0,0,0,1,1H30.72a1,1,0,0,0,1-1V10a1,1,0,0,0-1-1H19.28a1,1,0,0,0-1,1V25Z"
              />
            </svg>
            Files
          </li>
          <li>
            <svg
              viewBox="0 0 50 35"
            >
              <path
                d="M26.09,21a1.14,1.14,0,0,1-.29.9,1.44,1.44,0,0,1-.94.25,1.28,1.28,0,0,1-1-.28A1.26,1.26,0,0,1,23.64,21a1.22,1.22,0,0,1,.27-0.89,1.32,1.32,0,0,1,.95-0.27,1.38,1.38,0,0,1,1,.26A1.22,1.22,0,0,1,26.09,21Zm-0.86-2.27H24.51A0.47,0.47,0,0,1,24,18.27l-0.22-6.18a0.47,0.47,0,0,1,.47-0.48h1.14a0.47,0.47,0,0,1,.47.48l-0.2,6.18A0.47,0.47,0,0,1,25.23,18.72Z"
              />
              <path
                d="M25,7.5a10,10,0,1,0,10,10A10,10,0,0,0,25,7.5Zm0,18c-3.34,0-8-2.88-8-8,0-3.47,3.1-8,8-8,3.32,0,8,2.86,8,8A8.16,8.16,0,0,1,25,25.5Z"
              />
            </svg>
            Issues
          </li>
          <li>
            <svg
              viewBox="0 0 50 35"
            >
              <path
                d="M30.42,16.57a3,3,0,0,1-1-1.07H19.75a3,3,0,0,1-1,1.07,2.77,2.77,0,0,1-1.48.4,2.93,2.93,0,1,1,0-5.87,2.77,2.77,0,0,1,1.48.4,3,3,0,0,1,1,1.07h9.63a3,3,0,0,1,1-1.07,2.77,2.77,0,0,1,1.48-.4,2.93,2.93,0,1,1,0,5.87A2.77,2.77,0,0,1,30.42,16.57Zm0.24-1.29a1.74,1.74,0,1,0,0-2.47A1.72,1.72,0,0,0,30.66,15.27Zm-12.19,0a1.74,1.74,0,1,0-2.47,0A1.72,1.72,0,0,0,18.47,15.27Zm0.24,13a2.77,2.77,0,0,1-1.48.4,2.93,2.93,0,1,1,0-5.87,2.77,2.77,0,0,1,1.48.4,3,3,0,0,1,1,1.07H29a1.58,1.58,0,0,0,1.47-1.47V21.5a0.13,0.13,0,0,0-.13-0.13H28.49a0.41,0.41,0,0,1-.29-0.7l3.38-3.38a0.46,0.46,0,0,1,.65,0l3.36,3.36a0.42,0.42,0,0,1-.3.72H33.52a0.16,0.16,0,0,0-.16.16v1.31a4.73,4.73,0,0,1-4.4,4.4H19.75A3,3,0,0,1,18.71,28.3ZM16,27a1.74,1.74,0,1,0,0-2.48A1.72,1.72,0,0,0,16,27Z"
              />
            </svg>
            Pull Requests
          </li>
          <li>
            <svg
              class="svg aside-icon replaced-svg"
              fill="#7E90B2"
              viewBox="0 0 107 75"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M53.5 60.6c-.8 0-1.5-.2-2.2-.5-.1 0-.5-.2-1-.5s-1.3-.7-2.5-1.4-2.3-1.4-3.4-2.2c-1.2-.9-2.4-1.9-3.6-2.9-1.3-1.1-2.4-2.4-3.5-3.7-1.1-1.4-1.9-3-2.6-4.6-.7-1.8-1.1-3.7-1.1-5.6V19.4c0-1.3.5-2.6 1.5-3.5.9-.9 2.2-1.4 3.5-1.4h29.6c2.7 0 4.9 2.2 4.9 4.9v19.8c0 1.9-.4 3.8-1.1 5.6-.6 1.7-1.5 3.2-2.6 4.6-1 1.4-2.2 2.6-3.5 3.7-1.1 1-2.3 2-3.6 2.9-1.1.8-2.2 1.5-3.4 2.2s-2 1.1-2.5 1.4l-1.2.6c-.3.2-1 .4-1.7.4zm-.7-6.4l-.2.1.2-.1zm1.3 0zM40.3 21v18.2c0 1.1.2 2.2.6 3.2.4 1.1 1 2.1 1.7 3.1.8 1 1.7 2 2.6 2.8 1 .9 2 1.7 3 2.4.9.6 1.9 1.3 2.9 1.8s1.7 1 2.1 1.2l.3.2.3-.2c.4-.2 1.1-.6 2.1-1.2s2-1.2 2.9-1.8c1-.7 2-1.5 3-2.4s1.8-1.8 2.6-2.8c.7-1 1.3-2 1.7-3.1.4-1 .6-2.1.6-3.2V21H40.3z"
              />
            </svg>
            Security
          </li>
        </ul>
      </aside>
      <div
        class="css-1xwln0l"
      >
        <div
          class="css-odz3mk"
        >
          <div
            class="css-1qpwtic"
          >
            <div
              class="css-1fnlmjy"
            >
              <div>
                Commits list
              </div>
            </div>
            <div
              class="css-1fefcv5"
            >
              <div
                class="css-g6t9o0"
              >
                Your commits will show up here
              </div>
              <table>
                <thead />
                <tbody>
                  <tr>
                    <td
                      class="css-be2ift"
                    >
                      <div
                        class="css-14zx4vh"
                      >
                        <div
                          class="css-1vm6a6x"
                        />
                        <div
                          class="css-j1vr76"
                        />
                      </div>
                    </td>
                    <td
                      class="css-be2ift"
                    >
                      <div
                        class="css-j1vr76"
                      />
                    </td>
                    <td
                      class="css-be2ift"
                    >
                      <div
                        class="css-j1vr76"
                      />
                    </td>
                    <td
                      class="css-be2ift"
                    >
                      <div
                        class="css-j1vr76"
                      />
                    </td>
                  </tr>
                  <tr>
                    <td
                      class="css-be2ift"
                    >
                      <div
                        class="css-14zx4vh"
                      >
                        <div
                          class="css-1vm6a6x"
                        />
                        <div
                          class="css-j1vr76"
                        />
                      </div>
                    </td>
                    <td
                      class="css-be2ift"
                    >
                      <div
                        class="css-j1vr76"
                      />
                    </td>
                    <td
                      class="css-be2ift"
                    >
                      <div
                        class="css-j1vr76"
                      />
                    </td>
                    <td
                      class="css-be2ift"
                    >
                      <div
                        class="css-j1vr76"
                      />
                    </td>
                  </tr>
                  <tr>
                    <td
                      class="css-be2ift"
                    >
                      <div
                        class="css-14zx4vh"
                      >
                        <div
                          class="css-1vm6a6x"
                        />
                        <div
                          class="css-j1vr76"
                        />
                      </div>
                    </td>
                    <td
                      class="css-be2ift"
                    >
                      <div
                        class="css-j1vr76"
                      />
                    </td>
                    <td
                      class="css-be2ift"
                    >
                      <div
                        class="css-j1vr76"
                      />
                    </td>
                    <td
                      class="css-be2ift"
                    >
                      <div
                        class="css-j1vr76"
                      />
                    </td>
                  </tr>
                  <tr>
                    <td
                      class="css-be2ift"
                    >
                      <div
                        class="css-14zx4vh"
                      >
                        <div
                          class="css-1vm6a6x"
                        />
                        <div
                          class="css-j1vr76"
                        />
                      </div>
                    </td>
                    <td
                      class="css-be2ift"
                    >
                      <div
                        class="css-j1vr76"
                      />
                    </td>
                    <td
                      class="css-be2ift"
                    >
                      <div
                        class="css-j1vr76"
                      />
                    </td>
                    <td
                      class="css-be2ift"
                    >
                      <div
                        class="css-j1vr76"
                      />
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
        <footer
          class="css-4xnkmh"
        >
          @ 2019 codacy - Automated code review
        </footer>
      </div>
    </main>
  </div>
</DocumentFragment>
`;
